File: packages/core-php/src/ClientData.php
<?php

declare(strict_types=1);

namespace OpenPasskey;

class ClientData
{
    public static function verify(
        string $clientDataJSONB64,
        string $expectedType,
        string $expectedChallenge,
        string $expectedOrigin,
    ): string {
        $raw = Base64Url::decode($clientDataJSONB64);
        $cd = json_decode($raw, true);
        if ($cd === null) {
            throw new WebAuthnException('type_mismatch', 'Invalid clientDataJSON');
        }

        if (($cd['type'] ?? '') !== $expectedType) {
            throw new WebAuthnException('type_mismatch');
        }
        if (($cd['challenge'] ?? '') !== $expectedChallenge) {
            throw new WebAuthnException('challenge_mismatch');
        }
        if (($cd['origin'] ?? '') !== $expectedOrigin) {
            throw new WebAuthnException('origin_mismatch');
        }
        if (isset($cd['tokenBinding']['status']) && $cd['tokenBinding']['status'] === 'present') {
            throw new WebAuthnException('token_binding_unsupported');
        }

        return $raw;
    }
}

File: packages/core-ruby/lib/open_passkey/client_data.rb
# frozen_string_literal: true

require "json"

module OpenPasskey
  module ClientData
    def self.verify(client_data_json_b64, expected_type, expected_challenge, expected_origin)
      raw = Base64Url.decode(client_data_json_b64)
      cd = JSON.parse(raw)

      unless cd["type"] == expected_type
        raise WebAuthnError.new("type_mismatch")
      end
      unless cd["challenge"] == expected_challenge
        raise WebAuthnError.new("challenge_mismatch")
      end
      unless cd["origin"] == expected_origin
        raise WebAuthnError.new("origin_mismatch")
      end
      if cd.dig("tokenBinding", "status") == "present"
        raise WebAuthnError.new("token_binding_unsupported")
      end

      raw
    rescue JSON::ParserError
      raise WebAuthnError.new("type_mismatch", "Invalid clientDataJSON")
    end
  end
end
